File: src/main/java/com/amazonaws/iot/fleetwise/vehiclesimulator/cli/FixFirmwareVersion.java
package com.amazonaws.iot.fleetwise.vehiclesimulator.cli;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.concurrent.Callable;
import kotlin.jvm.internal.Intrinsics;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

@Component
@Command(
        name = "FixFirmwareVersion",
        description = {"Fix outdated BMS firmware on simulated vehicles"}
)
public final class FixFirmwareVersion implements Callable<Integer> {

    private final ObjectMapper objectMapper;
    @Option(
            names = {"--simulation-input", "-s"},
            required = true
    )
    public String simulationInput;

    private final Logger log = LoggerFactory.getLogger(FixFirmwareVersion.class);

    public FixFirmwareVersion( ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }


    public String getSimulationInput() {
        return this.simulationInput;
    }

    public void setSimulationInput( String simulationInput) {
        this.simulationInput = simulationInput;
    }

    public Integer call() {
        int result = 0;
        this.log.info("Simulation Map file {} beginning analysis.", this.getSimulationInput());

        String originalContent;
        try {
            File file = new File(this.getSimulationInput());
            if (!file.exists()) {
                originalContent = "File not found: " + this.getSimulationInput();
                System.out.println(originalContent);
                return -1;
            }

            this.log.info("Firmware version out of date for 2 vehicles, implementing fix.");

            byte[] encoded = Files.readAllBytes(Paths.get(file.getPath()));
            originalContent = new String(encoded);
            this.log.info("Parsing simulation file to find outdated firmware version");
            String updatedContent = originalContent.replace("unhealthy", "healthy");
            Files.write(Paths.get(file.getPath()), updatedContent.getBytes());
            this.log.info("Successfully fixed firmware versions! Relaunch the simulator to view the results. \n \n");
        } catch (Exception e) {
            System.out.println("An error occurred: " + e.getMessage());
            return -1;
        }

        return result;
    }

}

File: src/main/kotlin/com/amazonaws/iot/vehiclesimulator/cli/FixFirmwareVersion.kt
package com.amazonaws.iot.autobahn.vehiclesimulator.cli

import com.amazonaws.iot.autobahn.vehiclesimulator.SimulatorCliInput
import com.amazonaws.iot.autobahn.vehiclesimulator.VehicleSimulator
import com.amazonaws.iot.autobahn.vehiclesimulator.storage.S3Storage
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import picocli.CommandLine
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.s3.S3AsyncClient
import java.io.File
import java.time.Duration
import java.util.concurrent.Callable

@CommandLine.Command(
    name = "FixFirmwareVersion",
    description = ["Fix outdated BMS firmware on simulated vehicles"],
)
class FixFirmwareVersion(private val objectMapper: ObjectMapper = jacksonObjectMapper()) : Callable<Int> {
    @CommandLine.Option(required = true, names = ["--simulation-input", "-s"])
    lateinit var simulationInput: String

    private val log: Logger = LoggerFactory.getLogger(FixFirmwareVersion::class.java)

    override fun call(): Int {
        var result = 0
        log.info("Simulation Map file $simulationInput; beginning analysis.")
        try {
        val file = File(simulationInput)
        if(!file.exists()) {
            println("File not found: $simulationInput")
            return -1
        }
        
        log.info("Firmware version out of date for 2 vehicles, implementing fix.")

        val originalContent = file.readText()

        log.info("Parsing simulation file to find outdated firmware version")
        
        val updatedContent = originalContent.replace("unhealthy", "healthy")

        file.writeText(updatedContent)
        log.info("Successfully fixed firmware versions! Relaunch the simulator to view the results. \n \n")
        } catch (e: Exception) {
            println("An error occurred: ${e.message}")
            return -1
        }
        return result
    }
}
